Add section on python development (#172)
merely-useful/merely-useful.github.io: py-development.Rmd

```{python}
sum_two_numbers(2, 5)
```

The returned value can be assigned to a variable:

```{python}
number_sum = sum_two_numbers(2, 5)
number_sum
```

A more versatile function could add any amount of numbers together and return
their sum:

```{python}
def sum_all_numbers(list_of_numbers):
    number_sum = 0
    for number in list_of_numbers:
        number_sum += number
    return number_sum

sum_all_numbers([1, 2, 3,])
```

A function can also return multiple outputs, e.g. we can return the number of
elements in addition to their sum:

```{python}
def sum_and_len_all_numbers(list_of_numbers):
    number_sum = 0
    number_len = 0
    for number in list_of_numbers:
        number_sum += number
        number_len += 1
    return number_sum, number_len
```

To capture the output, we can either assign to a single name, a tuple, or
assign both at the same time to different variables.

```{python}
sum_and_len_of_numbers = sum_and_len_all_numbers([1, 2, 3])
sum_and_len_of_numbers
```

```{python}
sum_of_numbers, len_of_numbers = sum_and_len_all_numbers([1, 2, 3])
sum_of_numbers
```

```{python}
len_of_numbers
```

> > Note: When we defined the function `sum_two_numbers()`, we referred to `num1`

Author: This is a suggestion for how to include more advanced tips. We should probably make a nicer div (maybe bootstrap like green or blue color), but looking to get feedback on the idea before settling on design.

Reviewer: Yea, not sure how we can do that across output document types (pdf vs ebook vs html). If you have a solid idea, let me know!

Also, I like this parahraph :+1:

Author: [This seems promising to me](https://stackoverflow.com/a/47294235/2166823), what do you think? We could create a function for "Advanced notes/Deep dives" (and maybe also "Exercises" if we want that to be more than a standard heading) that authors could call when inserting a similar paragraph as mine here.

Although this now appears technically feasible, and I like having a visually distinct element for advanced notes, it appears I am the only one who has written in this style so far (the rest follow your previous suggestion of having everything in paragraphs. So I am inclined to wander down the path of least resistance here and change my writing instead of encouraging everyone else to alter theirs.

Reviewer: We can revisit this later. That link looks promising..

```{python}
def sum_and_len(list_of_numbers):
    return sum(list_of_numbers), len(list_of_numbers)
```

#### Defining default values {py-defining-default-values}

Up until now, our function calls have included enough arguments to assign one
to each of the function parameters. This can get tedious in functions with many
parameters and it is often beneficial to define default values for several of
the function parameters. As an example, we can modify `sum_two_numbers()` to
optionally return the two input arguments.

```{python}
def sum_two_numbers(num1, num2, return_input=False):
    if return_input:
        return num1, num2, num1 + num2
    else:
        return num1 + num2
```

By default the function will work just as previously.

```{python}
sum_two_numbers(2, 5)
```

But we now also have the option to return the input numbers.

```{python}
sum_two_numbers(2, 5, return_input=True)
```

Since the arguments are given in the same order as in the function definition,
we could have left out the `return_input=` part and just written `True` in the
third position. If we want to supply the arguments out of order, we need to
specify the parameter name of each argument.

> #### Function documentation {py-function-documentation}

Author: I am not including the Python type annotations. I guess that would be best practice since it allows IDEs to autodetect easier, but I find it quite messy with default arguments (and I believe many IDEs can read numpy doscstring for types).

```{python}
def sum_two_numbers(num1, num2):
    '''
    Add two numbers

    Args:
        num1 (int, float): The first number to be added.
        num2 (int, float): The second number to be added.

    Returns:
        int, float: Sum of the two numbers.
    '''

    return num1 + num2
```

The above style is commonly referred to as the Google docstring convention. As
functions get longer it can be beneficial to adapt the more elaborate `numpy`
docstring convention used in many scientific Python packages:

```{python}
def sum_two_numbers(num1, num2):
    '''
    Add two numbers

    Parameters
    ----------
    num1: int, float
        The first number to be added.
    num2: int, float
        The second number to be added.

    Returns
    -------
    int, float
        Sum of the two numbers.
    '''

    return num1 + num2
```

We recommend using the `numpy` docstring format since this is easy to read
complex function with many arguments and heavily used by the data science
Python packages. Docstring are what show up in the function help messages, so
it is important that these are well-written and helpful for the user.

```{python}
help(sum_two_numbers)
```

> ### Applying what we have learnt {py-applying-what-we-have-learnt}

Author: It would be great if they have written some small script previously in the course that we could modularize into functions here.

```{python error=TRUE}
assert 1 == 0
print('This is only printed if the assertion above is "True"')
```

Since the condition above is `False`, an `AssertionError` is raised and code
execution halts. If the assertion is `True`, there is no output and the next
line of code is executed (if there is one).

```{python}
assert 0 == 0
print('This is only printed if the assertion above is "True"')
```

As mentioned above, this is useful if we know that a variable should look a
certain way, since we can assert if this is the case and guard ourselves from
errors that origin early in the pipeline but could give rise to more cryptic
errors that are difficult to troubleshoot later in the pipeline.

It is often helpful to add a clarifying message to the assertion statement,
especially as assertions become more complex.

```{python error=TRUE}
x = 1
assert x == 0, 'x is not 0'
```

> Note: Similar functionality can be achieved by explicitly raising an error
> within a `try` and `except` block, or within a conditional statement using
> `if`, `elif`, and `else`, but `assert` is a simple and readable way of
> allowing for manual error checking.

## Packages {py-packages}

> ### Installing Python {py-installing-Python}

Author: The following two sections are related to packages and especially package managers so I included them here for now.

sum_two_numbers(2, 5, return_input=True)
```

Since the arguments are given in the same order as in the function definition,
we could have left out the `return_input=` part and just written `True` in the
third position. If we want to supply the arguments out of order, we need to
specify the parameter name of each argument.

#### Function documentation {#py-dev-function-documentation}

Functions might appear self-explanatory when they are being written, but it is
essential that there is proper documentation describing what the function does
and what types of arguments should be in the input. This helps other people who
are reading your code and also your future self that will be reusing these
functions.

In Python, a function is documented in its docstring, which is a multiline
Python string immediately following the function definition. It is surrounded
by triple quotes (either single or double) and can look like this.

```{python}
def sum_two_numbers(num1, num2):
    '''
    Add two numbers

    Args:
        num1 (int, float): The first number to be added.
        num2 (int, float): The second number to be added.

    Returns:
        int, float: Sum of the two numbers.

>     '''

Reviewer: Is there an easy way to add this docstring? E.g. like a key binding (for instance, Ctrl-Shift-R or whatever in RStudio for using roxygen comments).

Author: Yes, good point I will add this. There is actually a keybinding, a menu item, and an autocomplete template ([gif here](https://github.com/spyder-ide/spyder/pull/8700))!

    return num1 + num2
```

The above style is commonly referred to as the Google docstring convention. As
functions get longer it can be beneficial to adapt the more elaborate `numpy`
docstring convention used in many scientific Python packages:

```{python}
def sum_two_numbers(num1, num2):
    '''
    Add two numbers

    Parameters
    ----------
    num1: int, float
        The first number to be added.
    num2: int, float
        The second number to be added.

    Returns
    -------
    int, float
        Sum of the two numbers.

>     '''

Reviewer: Does this render into a nice format or this is how it looks in the help? Does indentation matter here? Does parameter order matter?

Author: Good point to bring up. Spyder has rich rendering of docstrings treating it similarly (or maybe exactly as) markdown. This means the NumPy-like docstrings work slightly better as the arguments are line separated without adding extra trailing spaces or newlines in the source:

Numpy docstring format:
![image](https://user-images.githubusercontent.com/4560057/64229334-5b188900-ce9e-11e9-94f0-23987f1d6302.png)

Google dostring format:
![image](https://user-images.githubusercontent.com/4560057/64229574-075a6f80-ce9f-11e9-83a6-a168102a2845.png)

I also prefer numpy docstring and they are the most common among scientific packages, so I think I will make a harder recommendation for Numpy format, exclude the Google example and replace it with a link, and add a screenshot of the rendered docstring.

In PyCharm, I believe both renders as rich text, in jupyterlab the raw string is shown.

Author: I believe the docstrings are parsed the same as markdown, so yes indentation will matter since it makes the parameter name and description render as a markdown definition list with the parameter name in bold typeface. The order doesn't matter for anything that I am aware of but it is convention to keep the order the same as in the function signature.

    return num1 + num2
```

We recommend using the `numpy` docstring format since this is easy to read
complex function with many arguments and heavily used by the data science
Python packages. Docstring are what show up in the function help messages, so
it is important that these are well-written and helpful for the user.

```{python}
help(sum_two_numbers)
```

### Applying what we have learnt {#py-dev-applying-what-we-have-learnt}

Now we can apply what we have learnt about functions on the code we have
written previously.

## How to make programs indicate that something has gone wrong?

> {#py-dev-how-to-make-programs-indicate-that-something-has-gone-wrong?}

Reviewer: This should be on the previous line

Author: I think these would actually work, but I changed them so it looks nicer.

When performing programmatic data analysis, we need to both ensure that our
code runs correctly and that it carries out the intended tasks. The Python
interpreter will help us with the first part; if a part of the code is not
valid, an error will be raised with a message that helps us trace back what
part of the code is not correct.

### Ducktyping - relying on Python to detect unexpected behavior {#py-dev-ducktyping

> --relying-on-Python-to-detect-unexpected-behavior}

Reviewer: This should be on the previous line.

In our function `sum_two_numbers()`, we could explicitly check that the input
arguments are numerical. However, explicitly checking the type of each input
parameter quickly becomes tedious and can make functions less readable,
especially as they grow more complex. An alternative approach is to try to
perform the intended operations on the input parameters and rely on that Python
will raise an error if they are of the wrong type.

```{python error=TRUE}
sum_two_numbers(5, 'six')
```

The raised error contains a helpful message that alerts us to what went wrong;
`string` objects cannot be added to `int` objects in Python. If the input
variables behave correctly then they probably are of the correct types and no
explicit checking is needed. This approach is often referred to as "ducktyping"

> because makes assumption on the type of variable based on its behavior, just

Reviewer: ```suggestion
because it makes assumptions on the type of variable based on its behavior, just
```

like the saying

> If it looks like a duck, swims like a duck, and quacks like a duck, then it
> probably is a duck.

### Assertions - explicitly checking for unexpected behavior {#py-dev-assertions--

> explicitly-checking-for-unexpected-behavior}

Reviewer: previous line

Ducktyping is useful to catch anything that raises an error in Python, but

> sometimes we might want to stop the code execution for other reason than a

Reviewer: ```suggestion
sometimes we might want to stop the code execution for reasons other than a
```

> technical Python error. Examples of this includes when you have performed a

Reviewer: ```suggestion
technical Python error. Examples of this include when you have performed a
```

specific operation that you know should give output of a certain shape and to
check if a variable is within an allowed range. The `assert` statement allows
us to check if a condition is `True` and stop code execution if it is not.

```{python error=TRUE}
assert 1 == 0
print('This is only printed if the assertion above is "True"')
```

Since the condition above is `False`, an `AssertionError` is raised and code
execution halts. If the assertion is `True`, there is no output and the next
line of code is executed (if there is one).

```{python}
assert 0 == 0
print('This is only printed if the assertion above is "True"')
```

As mentioned above, this is useful if we know that a variable should look a
certain way, since we can assert if this is the case and guard ourselves from

> errors that origin early in the pipeline but could give rise to more cryptic

Reviewer: ```suggestion
errors that originate early in the pipeline but could give rise to more cryptic
```

errors that are difficult to troubleshoot later in the pipeline.

It is often helpful to add a clarifying message to the assertion statement,
especially as assertions become more complex.

```{python error=TRUE}
x = 1
assert x == 0, 'x is not 0'
```

> Note: Similar functionality can be achieved by explicitly raising an error
> within a `try` and `except` block, or within a conditional statement using
> `if`, `elif`, and `else`, but `assert` is a simple and readable way of
> allowing for manual error checking.

## Packages {#py-dev-packages}

### Installing Python {#py-dev-installing-Python}

There are several ways of installing Python on your system. One of the most

> robust and cross-platform compatible is to install the Anaconda Python

Reviewer: robust what?

Author: buzzZZzword for SEO and to generate hYPeZ among readers.

I can replace with "reliable", "straightforward" or "hassle free", something along those lines.

includes graphical interfaces for interacting with Python, which can be invoked
by running `spyder` or `jupyter-lab` from the command line, these are covered
more in detail elsewhere in this book.

### What is a package {#py-dev-what-is-a-package}

Certain functionality that is considered essential for the Python programming
language is available wherever Python is installed. Other highly useful, but
often more domain specific functionality can be accessed separately in the
form of Python packages. A package is essentially a few Python scripts in a
specific directory structure coupled with installation instructions for the
computer.

> Note: You will sometimes hear packages refers to as "modules". The two
> words are often used interchangeably. Technically, a package is a folder that
> contains modules (scripts), which in turn contains functions (code).

People around the world have created packages for Python and made them freely
available for others to use resulting in one of the richest package ecosystems
for any programming language, with packages for web design, prose writing, game
development, and data science (to name just a few). Since there are so many

> packages available, it is infeasible to include all of them with the default

Reviewer: ```suggestion
packages available, it is not feasible to include all of them with the default
```

all available function and intelligently filter them as you type out more
letters, try it! When you start getting familiar with typing function names,
you will notice that this is often faster than looking for functions in menus.

It is common to give packages shorter nicknames, which are faster to type. This
is not necessary, but can save work in long files and make code less verbose so
that it is easier to read:

```{python}
import numpy as np

np.mean([1, 2, 3])
```

We could also import the mean function directly.

```{python}
from numpy import mean

mean([1, 2, 3])
```

And even give it a nickname.

```{python}
from numpy import mean as mn

mn([1, 2, 3])
```

Which of these you use is up to you, but it is common to follow the conventions
established by the library's authors, which for numpy is `import numpy as np`
and use functions via `np.<function_name>`. One thing to avoid is to import
everything from a package, e.g. `from numpy import *`. If this is done with
every package, it is almost guaranteed that the same function name will be
available from more than one package and it will be difficult to keep track of

> if you are using the `mean` function from `numpy` or from another package that

Reviewer: ```suggestion
it if you are using the `mean` function from `numpy` or from another package that
```

Python packages. For comprehensive instructions on both of these, refer to the
[official
documentation](https://docs.continuum.io/anaconda/#navigator-or-conda). Brief
step-by-step instructions to get up and running with `conda` follow.

1. To install a new Python package from the Anaconda repositories, simply run
   `conda install <package name>` in a terminal. You can also use the `pip`
   package manager, but it will be easier to keep track of packages by sticking
   to one installation method.
2. Some packages are not available in the default Anaconda repositories. User
   contributed packaged are available in Anaconda "channels", use `anaconda
   search -t conda <package name>`, to find a channel with the desired package.
   To install this package, use `conda install -c <channel name> <package
   name>`. The [conda forge channel](https://conda-forge.github.io/) channel
   has many of the packages not in the default repositories.

## How to get help online {#py-dev-how-to-get-help-online}

When reading the built-in Python help is not sufficient, there are several 
online resources that can helpful. One of the most commonly used resources for

> data science is the Stack Exchange network wich offer Q&A sites both for

Reviewer: ```suggestion
data science is the Stack Exchange network which offer Q&A sites both for
```

programming related topics via [Stack Overflow][so], as well as statistics and
machine learning via [Cross Validated][cv].

If nothing relevant can be found after searching the many existing question and
answers on these sites, it is appropriate to ask a new question! Stack Overflow
has detailed instruction on how to create a [minimal reproducible

> example][so-mre] when asking a question to increase the chances that the

Reviewer: In R there is the repro package to automatically create a MRE, rendered into the markdown form needed for SO. Is there something similar in Python?

Author: I had a brief look, but could not find anything.

question receives a specific and helpful answer. The key principles listed on

> the website recommends that and answer follows these guidelines: 

Reviewer: ```suggestion
the website recommends that an answer follows these guidelines: 
```

> - Minimal – Use as little code as possible that still produces the same
> problem
> - Complete – Provide all parts someone else needs to reproduce your problem
> in the question itself
> - Reproducible – Test the code you're about to provide to make sure it
> reproduces the problem

An additional benefit of reducing the problem into this format is that you
might discover the error in the process! This process helps narrowing down
exactly which region of the code is failing and the act of explaining the
problem often reveals a solution before any replies have come in (commonly
referred to as [rubber duck debugging][wiki-rubber-duck-debugging].

## Key Points {#py-dev-keypoints}

```{r child="keypoints/py-development.md"}
```

> ## Exercises {#py-dev-exercises}

Reviewer: It might help if these exercises either had more code scaffolding or were written in a way to achieve a specific output based on an input. As they are now, they are too vague.

Author: I have added explicit instruction to start from the functions I had in mind from the examples in the section text.

> 1. Write a function that returns the mean of exactly two numbers.

Reviewer: It might help to create some code scaffolding for this exercise.

Author: The scaffolding for this is in the chapter where we walk through how to create a function that calculates the sum of two numbers and the sum and length of an arbitrary long list of numbers. I can add a reminder to revisit that section.

2. Write a function that returns the mean of any amount of numbers.
3. Write function that takes a string as its input and returns the
   first and last character of the string.
4. Write a function that returns any character from the string. Which character
   is returned should be determined by an integer specified via a named
   parameter that has the default value of `0`.

> 5. Expand on the previous function by include a assert statement that checks if

Reviewer: If this is a multi-part exercise, make it one exercise with two parts, not two exercises that are linked.
